// textlint/lib/rule/compute-location.js
// LICENSE : MIT
"use strict";
/**
 * compute node's location with error's padding location.
 *
 * @param {TxtNode} node
 * @param {RuleError} padding
 * @returns {{line: number, column: number, fix?: FixCommand}}
 */
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports["default"] = computeLocation;

function computeLocation(node, padding) {
    var nodeRange = node.range;
    var line = node.loc.start.line;
    var column = node.loc.start.column;
    if (padding.line > 0) {
        line += padding.line;
        // when report with padding {line, column}, message.column should be 0 + padding.column.
        // In other word, padding line > 0 and message.column start with 0.
        if (padding.column) {
            // means 0 + padding column
            column = padding.column;
        }
    } else {
        if (padding.column) {
            column += padding.column;
        }
    }
    // fix(command) is relative from node's range
    if (padding.fix) {
        var fix = {
            range: [nodeRange[0] + padding.fix.range[0], nodeRange[0] + padding.fix.range[1]],
            text: padding.fix.text
        };
        return {
            line: line,
            column: column,
            fix: fix
        };
    }
    return {
        line: line,
        column: column
    };
}

module.exports = exports["default"];
